src/Microsoft.AspNetCore.Blazor/Components/ParameterCollectionExtensions.cs

```diff
@@ -26,47 +38,89 @@ public static class ParameterCollectionExtensions
+            }
+
             foreach (var parameter in parameterCollection)
             {
-                AssignToProperty(target, parameter);
+                var parameterName = parameter.Name;
+                if (!parameterWriters.TryGetValue(parameterName, out var parameterWriter))
+                {
+                    ThrowForUnknownIncomingParameterName(targetType, parameterName);
+                }
+
+                try
+                {
+                    parameterWriter(ref parameter.Frame, target);
+                }
+                catch (Exception ex)
+                {
+                    throw new InvalidOperationException(
+                        $"Unable to set property '{parameterName}' on object of " +
+                        $"type '{target.GetType().FullName}'. The error was: {ex.Message}", ex);
+                }
             }
         }
 
-        private static void AssignToProperty(object target, Parameter parameter)
+        private static IDictionary<string, WriteParameterAction> CreateParameterWriters(Type targetType)
         {
-            // TODO: Don't just use naive reflection like this. Possible ways to make it faster:
-            // (a) Create and cache a property-assigning open delegate for each (target type,
-            //     property name) pair, e.g., using propertyInfo.GetSetMethod().CreateDelegate(...)
-            //     That's much faster than caching the PropertyInfo, at least on JIT-enabled platforms.
-            // (b) Or possibly just code-gen an IComponent.SetParameters implementation for each
-            //     Razor component. However that might not work well with code-behind inheritance,
-            //     because the code-behind wouldn't be able to override it.
+            var result = new Dictionary<string, WriteParameterAction>(StringComparer.OrdinalIgnoreCase);
 
-            var propertyInfo = GetPropertyInfo(target.GetType(), parameter.Name);
-            try
-            {
-                propertyInfo.SetValue(target, parameter.Value);
-            }
-            catch (Exception ex)
+            foreach (var propertyInfo in GetBindableProperties(targetType))
             {
-                throw new InvalidOperationException(
-                    $"Unable to set property '{parameter.Name}' on object of " +
-                    $"type '{target.GetType().FullName}'. The error was: {ex.Message}", ex);
+                var propertySetter = MemberAssignment.CreatePropertySetter(targetType, propertyInfo);
+
+                var propertyName = propertyInfo.Name;
+                if (result.ContainsKey(propertyName))
+                {
+                    throw new InvalidOperationException(
+                        $"The type '{targetType.FullName}' declares more than one parameter matching the " +
+                        $"name '{propertyName.ToLowerInvariant()}'. Parameter names are case-insensitive and must be unique.");
```

Comment:
I think the other case you could hit here is property visibility. Using `new` to hide a property will trigger this error I think because `MemberAssignment.GetBindableProperties` doesn't account for that.

Reply:
Yes, that's deliberately the case. We treat it as an error to shadow a property that has `[Parameter]`, because if you do that there's no way for anyone to set the shadowed parameter, and yet the inherited class can legitimately require it to be set. There's a unit test for that scenario too.